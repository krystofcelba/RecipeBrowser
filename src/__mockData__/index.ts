export const base64Uri = 'data:image/jpeg;base64,xxxxxxxxx';

export const addRecipeFormValues = {
  name: 'recipe',
  description: 'desc',
  image: base64Uri,
  ingredients: [{ name: 'aa', amount: 1, unit: 'kg' }] as [{ name: string; amount: number; unit: string }],
  seasonings: ['red powder'] as [string],
  steps: ['ddd'] as [string],
};

export const mockedImagePickerImage = {
  path: '/xxx',
  mime: 'image/json',
  data: 'xxxxxxx',
};
