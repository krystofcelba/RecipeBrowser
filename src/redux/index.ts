import { applyMiddleware, createStore, compose, Store } from 'redux';
import { composeWithDevTools } from 'remote-redux-devtools';
import createSagaMiddleware from 'redux-saga';

import rootReducer, { RootState } from './reducers';
import rootSaga from './sagas';

const configureStore = (): Store<RootState> => {
  const sagaMiddleware = createSagaMiddleware();

  const composeEnhancers = __DEV__
    ? composeWithDevTools({
        realtime: true,
        hostname: 'localhost',
        port: 8000,
      })
    : compose;

  const store = createStore(
    rootReducer,
    composeEnhancers(applyMiddleware(sagaMiddleware)),
  );

  sagaMiddleware.run(rootSaga);
  return store;
};

export default configureStore;
