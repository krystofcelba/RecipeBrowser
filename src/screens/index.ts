import { Navigation } from 'react-native-navigation';

import RecipesList from './RecipesList';

export const registerScreens = (store, provider) => {
  Navigation.registerComponent(
    'recipeBrowser.RecipesList',
    () => RecipesList,
    store,
    provider,
  );
};
