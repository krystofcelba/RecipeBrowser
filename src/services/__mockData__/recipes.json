[
  {
    "description":
      "This buffet classic can also be healthy! Mayonnaise is replaced by cottage cheese. You can blend the filling for more creaminess.",
    "id": 9,
    "image": "https://nutrition-assets.freeletics.com/recipes/large/mustard_deviled_eggs.jpg",
    "ingredients": [
      {
        "amount": 4,
        "name": "hard-boiled eggs",
        "unit": ""
      },
      {
        "amount": 2,
        "name": "Dijon mustard",
        "unit": "tsp"
      },
      {
        "amount": 100,
        "name": "cottage cheese",
        "unit": "g"
      },
      {
        "amount": 2,
        "name": "chopped dill",
        "unit": "tbsp"
      }
    ],
    "name": "Mustard Deviled Eggs",
    "seasonings": ["Paprika powder"],
    "steps": [
      "Halve eggs and remove yolk.",
      "Mix yolks with mustard, cottage cheese and half of dill.",
      "Fill egg halves with yolk mixture and garnish with remaining dill."
    ]
  }
]
