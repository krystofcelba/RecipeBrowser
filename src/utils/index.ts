import i18n from 'src/assets/i18n';
import { Map } from 'immutable';

export const isString = value => typeof value === 'string';
export const isNumber = value => !isNaN(Number(value));

export const formatImageBase64Uri = ({ mime, data }: { mime: string; data: string }) => `data:${mime};base64,${data}`;
export const extractMimeType = (base64Uri: string) => base64Uri.slice(5, base64Uri.indexOf(';base64,'));

export const validateNewRecipeFormValues = (values: Map<string, any>) => {
  let errors = {};
  const { name, description, ingredients, seasonings, steps } = values.toJS();
  if (!name) {
    errors = { ...errors, name: i18n.t('requiredError', { field: i18n.t('name') }) };
  }
  if (!description) {
    errors = { ...errors, description: i18n.t('requiredError', { field: i18n.t('description') }) };
  }

  if (!ingredients) {
    errors = {
      ...errors,
      ingredients: { _error: i18n.t('atLeastOneError', { field: i18n.t('ingredient').toLowerCase() }) },
    };
  } else {
    const ingredientsErrors = ingredients.map(value => {
      const { name, amount, unit } = value;
      let err = '';
      if (!name) {
        err = `${i18n.t('requiredError', { field: i18n.t('name') })}. `;
      }
      if (!amount) {
        err = `${err}${i18n.t('requiredError', { field: i18n.t('amount') })}. `;
      } else if (amount && !isNumber(amount)) {
        err = `${err}${i18n.t('mustBeNumberError', { field: i18n.t('amount') })}. `;
      }
      if (!unit) {
        err = `${err}${i18n.t('requiredError', { field: i18n.t('unit') })}. `;
      }
      return err;
    });
    errors = { ...errors, ingredients: ingredientsErrors };
  }

  if (!seasonings) {
    errors = {
      ...errors,
      seasonings: { _error: i18n.t('atLeastOneError', { field: i18n.t('seasoning').toLowerCase() }) },
    };
  } else {
    const arrayErrors = seasonings.map(value => {
      return !isString(value) ? `${i18n.t('requiredError', { field: i18n.t('value') })}. ` : '';
    });
    errors = { ...errors, seasonings: arrayErrors };
  }

  if (!steps || !steps.length) {
    errors = {
      ...errors,
      steps: { _error: i18n.t('atLeastOneError', { field: i18n.t('step').toLowerCase() }) },
    };
  } else {
    const arrayErrors = steps.map(value => {
      return !isString(value) ? `${i18n.t('requiredError', { field: i18n.t('value') })}. ` : '';
    });
    errors = { ...errors, steps: arrayErrors };
  }

  return errors;
};
