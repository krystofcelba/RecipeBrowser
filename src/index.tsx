import { Component } from 'react';
import { Provider } from 'react-redux';
import { Navigation } from 'react-native-navigation';

import configureStore from './redux';
import { registerScreens } from './screens';
import { navigatorStyle } from './assets/constants';

const store = configureStore();
registerScreens(store, Provider);

class App extends Component {
  constructor(props) {
    super(props);
    this.startApp();
  }

  startApp = () => {
    Navigation.startSingleScreenApp({
      screen: {
        screen: 'recipeBrowser.RecipesList',
        ...navigatorStyle,
      },
      appStyle: {
        keepStyleAcrossPush: false,
      },
    });
  };
}

export default App;
