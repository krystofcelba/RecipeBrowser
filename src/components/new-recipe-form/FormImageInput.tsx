import React from 'react';
import { StyleSheet, StyleProp, TextStyle, TouchableOpacity, Image } from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import ImagePicker from 'react-native-image-crop-picker';

import { imagePickerImageSize } from 'src/assets/constants';
import { formatImageBase64Uri } from 'src/utils';

interface Props {
  input: {
    name: string;
    value: string;
    onChange: (text: string) => void;
  };
}

class FormImageInput extends React.Component<Props> {
  onPress = async () => {
    try {
      const { input: { onChange } } = this.props;
      const image: any = await ImagePicker.openPicker({
        ...imagePickerImageSize,
        cropping: true,
        includeBase64: true,
      });
      onChange(formatImageBase64Uri(image));
    } catch (error) {}
  };

  render() {
    const { input: { value } } = this.props;
    return (
      <TouchableOpacity style={styles.imageContainer} onPress={this.onPress}>
        {value.length ? (
          <Image style={styles.recipeImage} source={{ uri: value }} resizeMode="cover" />
        ) : (
          <Icon name="ios-images-outline" size={100} />
        )}
      </TouchableOpacity>
    );
  }
}

const styles = StyleSheet.create({
  imageContainer: {
    width: '100%',
    height: 220,
    alignItems: 'center',
    justifyContent: 'center',
  },
  recipeImage: {
    width: '100%',
    height: '100%',
  },
});

export default FormImageInput;
