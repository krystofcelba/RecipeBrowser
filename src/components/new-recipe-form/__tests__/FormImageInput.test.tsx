import React from 'react';
import { shallow } from 'enzyme';

import FormImageInput from '../FormImageInput';
import { base64Uri } from 'src/__mockData__';

describe('FormImageInput component', () => {
  it('renders as expected', () => {
    const wrapper = shallow(<FormImageInput input={{ name: 'image', value: '', onChange: () => {} }} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as expected with image', async () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <FormImageInput
        input={{
          name: 'image',
          value: base64Uri,
          onChange,
        }}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onChange as expected', async () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <FormImageInput
        input={{
          name: '',
          value: '',
          onChange,
        }}
      />,
    );
    await wrapper.instance().onPress();
    expect(onChange.mock.calls).toMatchSnapshot();
  });
});
