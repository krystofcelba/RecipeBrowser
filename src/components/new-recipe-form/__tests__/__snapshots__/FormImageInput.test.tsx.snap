// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormImageInput component calls onChange as expected 1`] = `
Array [
  Array [
    "data:image/json;base64,xxxxxxx",
  ],
]
`;

exports[`FormImageInput component renders as expected 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "height": 220,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Icon
    allowFontScaling={false}
    name="ios-images-outline"
    size={100}
  />
</TouchableOpacity>
`;

exports[`FormImageInput component renders as expected with image 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "height": 220,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "data:image/jpeg;base64,xxxxxxxxx",
      }
    }
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
</TouchableOpacity>
`;
